File: dart/jeliya_protocol/lib/src/conventions/join.dart
/// `room.join` with the reference retry ladder, ported 1:1 from
/// ui/src/lib/join.ts: 5 attempts, 1.5/2/3/4s delays, retrying ONLY
/// `peer_unreachable` (any other error — and the final attempt — rethrows
/// immediately).
library;

import '../methods.dart';
import '../models.dart';
import '../protocol.dart';

/// The two join-progress phases (join.ts `JoinPhase`).
abstract final class JoinPhases {
  static const String connecting = 'connecting';
  static const String retrying = 'retrying';
  static const List<String> all = [connecting, retrying];
}

/// One progress callback payload (join.ts `JoinProgress`).
class JoinProgress {
  const JoinProgress({
    required this.phase,
    required this.attempt,
    required this.maxAttempts,
    required this.message,
    this.lastError,
  });

  /// One of [JoinPhases.all].
  final String phase;
  final int attempt;
  final int maxAttempts;

  /// Reference UI copy, emitted verbatim so both clients narrate joins the
  /// same way.
  final String message;

  /// The shaped `peer_unreachable` error that triggered a retry; null on
  /// [JoinPhases.connecting].
  final RequestError? lastError;
}

const int _joinAttempts = 5;
const List<Duration> _retryDelays = [
  Duration(milliseconds: 1500),
  Duration(milliseconds: 2000),
  Duration(milliseconds: 3000),
  Duration(milliseconds: 4000),
];

/// Join a room, retrying only [ErrorCodes.peerUnreachable] (the inviter's
/// first path did not answer) up to 5 attempts with 1.5/2/3/4s delays.
/// Returns the joined `room_id`; any other failure — or the fifth
/// `peer_unreachable` — rethrows the original error. [peers] are optional
/// `<endpoint_id>@<ip:port>` dial hints. [sleep] is a test seam and defaults
/// to a real [Future.delayed].
Future<String> joinRoomWithRetry(
  Client client, {
  required String ticket,
  List<String>? peers,
  void Function(JoinProgress progress)? onProgress,
  Future<void> Function(Duration delay)? sleep,
}) async {
  final wait = sleep ?? (delay) => Future<void>.delayed(delay);

  for (var attempt = 1; attempt <= _joinAttempts; attempt++) {
    onProgress?.call(JoinProgress(
      phase: JoinPhases.connecting,
      attempt: attempt,
      maxAttempts: _joinAttempts,
      message: attempt == 1
          ? 'Finding the inviter and syncing the room invite...'
          : 'Retrying join ($attempt/$_joinAttempts)...',
    ));

    try {
      return await client.roomJoin(ticket, peers: peers);
    } catch (e) {
      final err = errorShape(e);
      if (err.code != ErrorCodes.peerUnreachable || attempt == _joinAttempts) {
        rethrow;
      }
      final delay = _retryDelays[
          attempt - 1 < _retryDelays.length ? attempt - 1 : _retryDelays.length - 1];
      onProgress?.call(JoinProgress(
        phase: JoinPhases.retrying,
        attempt: attempt,
        maxAttempts: _joinAttempts,
        message: 'The first path did not answer. '
            'Retrying in ${(delay.inMilliseconds / 1000).round()}s...',
        lastError: err,
      ));
      await wait(delay);
    }
  }

  // Unreachable: the fifth attempt either returned or rethrew above (the
  // reference's trailing `throw lastError` is the same dead end).
  throw StateError('joinRoomWithRetry: exhausted attempts without a rethrow');
}

File: ui/src/lib/join.ts
import type { Client, DaemonErrorShape } from './protocol';
import { errorShape, RequestError } from './protocol';

export type JoinPhase = 'connecting' | 'retrying';

export interface JoinProgress {
  phase: JoinPhase;
  attempt: number;
  maxAttempts: number;
  message: string;
  lastError?: DaemonErrorShape;
}

const JOIN_ATTEMPTS = 5;
const RETRY_DELAYS_MS = [1_500, 2_000, 3_000, 4_000];

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => window.setTimeout(resolve, ms));
}

export async function joinRoomWithRetry(
  client: Client,
  params: { ticket: string; peers?: string[] },
  onProgress?: (progress: JoinProgress) => void,
): Promise<{ room_id: string }> {
  let lastError: unknown = null;

  for (let attempt = 1; attempt <= JOIN_ATTEMPTS; attempt += 1) {
    onProgress?.({
      phase: 'connecting',
      attempt,
      maxAttempts: JOIN_ATTEMPTS,
      message:
        attempt === 1
          ? 'Finding the inviter and syncing the room invite...'
          : `Retrying join (${attempt}/${JOIN_ATTEMPTS})...`,
    });

    try {
      return await client.call('room.join', params);
    } catch (e) {
      lastError = e;
      const err = errorShape(e);
      if (err.code !== 'peer_unreachable' || attempt === JOIN_ATTEMPTS) {
        throw e;
      }
      const delay = RETRY_DELAYS_MS[Math.min(attempt - 1, RETRY_DELAYS_MS.length - 1)];
      onProgress?.({
        phase: 'retrying',
        attempt,
        maxAttempts: JOIN_ATTEMPTS,
        message: `The first path did not answer. Retrying in ${Math.round(delay / 1000)}s...`,
        lastError: err,
      });
      await sleep(delay);
      // The daemon connection can drop during the backoff. Issuing the next
      // attempt then would QUEUE it until reconnect (WsClient keeps unsent
      // requests), pinning the caller's busy state — and the contained join
      // dialog with it — for as long as the daemon stays down. Fail honestly
      // instead: the dialog surfaces the real error and dismissal returns.
      if (client.getState() !== 'connected') {
        throw new RequestError({
          code: 'connection_lost',
          message: 'the daemon connection dropped while waiting to retry the join',
          hint: 'reconnect, then paste the same invite again',
        });
      }
    }
  }

  throw lastError;
}
